src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography
+{
+    internal sealed partial class RandomNumberGeneratorImplementation : RandomNumberGenerator
+    {
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+
+            GetBytes(data);
+
+            int indexOfFirst0Byte = data.Length;
+            for (int i = 0; i < data.Length; i++)
+            {
+                if (data[i] == 0)
+                {
+                    indexOfFirst0Byte = i;
+                    break;
+                }
+            }
+
+            for (int i = indexOfFirst0Byte; i < data.Length; i++)
+            {
+                if (data[i] != 0)
+                {
+                    data[indexOfFirst0Byte++] = data[i];
+                }
+            }
+
+            while (indexOfFirst0Byte < data.Length)
+            {
+                // This should be more than enough to fill the rest in one iteration
+                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
```

Comment:
Agree, just pointing out perhaps some of the original intent. In your solution, there are zero allocations however perhaps an extra call to get random numbers for large buffers, which does have some overhead although probably not as much as extra allocs.